File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    void dfs(char[][] grid, int i, int j ){
        int totalRows = grid.length;
        int totalCols = grid[0].length;
        
        if(i<0 || j <0 || i >= totalRows || j >= totalCols ){
            return ;
        }
        else if(grid[i][j] != '1'){
            return;  // either water or land already visited 
        }
        
        grid[i][j] = '2'; // marking land as visited
        
        
        // left, right, top, down
        dfs(grid, i, j-1);
        dfs(grid, i, j+1);
        dfs(grid, i-1, j);
        dfs(grid, i+1, j);
        
        return ;
        
    }
    public int numIslands(char[][] grid) {
        // using dfs
        int n = grid.length;
        int m = grid[0].length;
        
        int  island = 0;
        for(int i = 0 ; i < n ; i++){
            for(int j = 0 ; j < m ; j++){
                if(grid[i][j] == '1'){
                    dfs(grid, i, j);
                    island++;
                }
            }
        }
        
        
        return island;
    }
}

File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    
    bool isvalid(int i, int j, int m, int n){
        if(i < 0 || j < 0 || i >= m || j >= n){
            return false;
        }
        
        return true;
    }
    
    void doDFS(vector<vector<char>> &grid, int i , int j){
        int m = grid.size();
        int n = grid[0].size();
        
        if(!isvalid(i,j,m,n)){
            return ;
        }
        
        if(grid[i][j] !=  '1'){
            // already visited ('2') or water ('0')
            return ;
        }
        
        grid[i][j] = '2';
        
        int dir[5] = {1,0,-1,0,1};
        
        for(int k = 0 ; k < 4 ; k++){
            int newx = i + dir[k];
            int newy = j + dir[k+1];
            doDFS(grid, newx, newy);
        }
        
        return ;
           
    }
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        
        for(int i = 0 ; i < grid.size() ; i++){
            for(int j = 0 ; j < grid[0].size() ; j++){
                if(grid[i][j] == '1'){
                    doDFS(grid, i, j);
                    count++;
                }
            }
        }
        
        return count;
        
    }
};